client/device/cloudftd/upgrade.go

```diff
@@ -116,3 +159,23 @@ func (f *ftdUpgradeService) validateUpgradePathExistsTo(ftdDevice *FtdDevice, to
 
 	return errors.New(fmt.Sprintf("%s is not a valid version to upgrade FTD device %s to", toVersion.String(), ftdDevice.Name))
 }
+
+func (f *ftdUpgradeService) getUpgradePackage(ftdDevice *FtdDevice, toVersion *ftd.Version) (*UpgradePackage, error) {
```

Comment:
I wouldn't duplicate the API call we're making in validateUpgradePathExistsTo here. Why don't you just change validateUpgradePathExistsTo to validateUpgradePathAndGetUpgradePackage, and change it to return *UpgradePackage, nil?

Reply:
And if softwareVersion.Equal(toVersion) return nil, nil, and then do not execute doUpgrade if upgradePackage is nil